### PR TITLE
[IMP] hr_leave: improve leave ux

### PR DESCRIPTION
**Pull request by Bertrand2**
Small changes in the layout of the hr_leave form view, mainly for the manager's view point

Task : #2734842 [Time Off] ux changes

### CODE REVIEW

File: addons/hr_holidays/views/resource_views.xml [hunk 1]
```diff
@@ -49,10 +49,10 @@
         <field name="res_model">resource.calendar.leaves</field>
         <field name="view_mode">tree</field>
         <field name="view_id" ref="resource_calendar_leaves_tree_inherit"/>
-        <field name="domain">[('resource_id', '=', False)]</field>
+        <field name="domain">['&amp;', ('resource_id', '=', False), ('calendar_id', 'in', (active_id, False))]</field>
         <field name="context">{
             'default_calendar_id': active_id,
-            'search_default_calendar_id': active_id,
+            'search_default_calendar_id': False,
```
Comment (kba-odoo) [inline addons/hr_holidays/views/resource_views.xml:55]: Is it not working if you simply remove the line?

File: addons/hr_holidays/models/resource.py [hunk 2]
```diff
@@ -1,10 +1,50 @@
...
+from odoo.exceptions import ValidationError
 
 
 class CalendarLeaves(models.Model):
     _inherit = "resource.calendar.leaves"
 
     holiday_id = fields.Many2one("hr.leave", string='Leave Request')
+
+    @api.constrains('date_from', 'date_to', 'calendar_id')
+    def check_compare_dates(self):
```
Comment (kba-odoo) [inline addons/hr_holidays/models/resource.py:14]: It's better to make it private:

```suggestion
    def _check_compare_dates(self):
```

File: addons/hr_holidays/models/resource.py [hunk 3]
```diff
@@ -1,10 +1,50 @@
...
+                    '&',
+                    ('date_from', '>', record.date_from),
+                    ('date_from', '<', record.date_to),
+                    '&',
+                    ('date_to', '>', record.date_from),
+                    ('date_to', '<', record.date_to),
+                    '&',
+                    ('date_from', '<=', record.date_from),
+                    ('date_to', '>=', record.date_to),
+                ])
```
Comment (kba-odoo) [inline addons/hr_holidays/models/resource.py:33]: Can't you do a single `search()` which will take all the leaves (using min(date_from)/max(date_to)) and then filter to see if you have any watching ones?

Here's an example of something similar: https://github.com/odoo/enterprise/blob/master/hr_payroll_holidays/models/hr_leave.py#L26 :-)

**Review by kbapt — CHANGES_REQUESTED** [review]

File: addons/hr_holidays/models/resource.py [hunk 4]
```diff
@@ -1,10 +1,41 @@
...
     _inherit = "resource.calendar.leaves"
 
     holiday_id = fields.Many2one("hr.leave", string='Leave Request')
+
+    @api.constrains('date_from', 'date_to', 'calendar_id')
+    def _check_compare_dates(self):
+        all_existing_leaves = self.env['resource.calendar.leaves'].search([
+            ('resource_id', '=', False),
+            ('date_from', '<=', max(self.mapped('date_to'))),
+            ('date_to', '>=', min(self.mapped('date_from'))),
```
Comment (kba-odoo) [inline addons/hr_holidays/models/resource.py:18]: The company should also be considered, especially if there are no calendar_id

**Review by kbapt — CHANGES_REQUESTED** [review]

File: addons/hr_holidays/models/resource.py [hunk 5]
```diff
@@ -1,10 +1,47 @@
...
 class CalendarLeaves(models.Model):
     _inherit = "resource.calendar.leaves"
 
     holiday_id = fields.Many2one("hr.leave", string='Leave Request')
+
+    @api.constrains('date_from', 'date_to', 'calendar_id')
+    def _check_compare_dates(self):
+        all_existing_leaves = self.env['resource.calendar.leaves'].search([
+            ('resource_id', '=', False),
+            ('company_id', 'in', self.mapped('company_id.id')),
```
Comment (kba-odoo) [inline addons/hr_holidays/models/resource.py:17]: You don't have to call `mapped()` on m2o, it's the default behavior when accessing them :-)
You can also use the attribute `.ids` to get the list of IDs of the recordset:

```suggestion
            ('company_id', 'in', self.company_id.ids),
```

**Review by kbapt — CHANGES_REQUESTED** [review]
Last comment :innocent:

**Comment by kba-odoo** [comment]
robodoo r+

**Pull request closed without merging**